### PR TITLE
Integrate deep learning object detection model for automatically labling 

### PR DESCRIPTION
**Pull request by vuthede**
I intergrate state-of-the-art deep learning object detection with Tracker to help us automatically labeling data. I create a new file for that named `main_auto.py`
We can test this usability.

### CODE REVIEW

**Comment by Cartucho** [comment]
Cool! Let's merge the other PR first

**Comment by Cartucho** [comment]
@vuthede why did you close this one? It wasn't working?

**Comment by Cartucho** [comment]
You can leave it open and then we squash the commits before merging. This way I can also help you out.

**Comment by vuthede** [comment]
Ok, i see, thanks. Btw, I am working on:
+ Fix bug.
+ Make it works with more classes.
+ Comment code.

**Comment by Cartucho** [comment]
I will have a look at that! Btw, we should compress the `people_walking.mp4` video, using [this website](http://www.videosmaller.com/) for example.

**Comment by Cartucho** [comment]
I didn't get a chance to look into this. However, this is by the far the most interesting PR that we have here!

Btw, what model are you using?

It would be nice to make this in a way to allow multiple models, we could have a folder called `main/automatic_labeling` and there a folder per model containing a `README` and a `main_auto.py`.

I also want to add a model for YOLOv3 and Mask R-CNN. 

So it could look like this:
&nbsp;&nbsp;&nbsp;&nbsp;`main/`
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; `input/`
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; `output/`
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; `automatic_labeling/`
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; `YOLOv3/`
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; `Mask_R-CNN/`

**Comment by vuthede** [comment]
I have used the pretrained models on that link https://github.com/tensorflow/models/blob/master/research/object_detection/g3doc/detection_model_zoo.md.
At the moment, i use **only one** file `main_auto.py` for using multiple model. In particular, I made the way so that we just change the path of pretrained model in file `main_auto.py` to use different model.  What do u think?

**Comment by Cartucho** [comment]
Wow, I didn't know! That sounds great, let's use a single file then!

If they choose the Mask R-CNN however we would need to be able to deal with pixel labeling (segmentation).

I will set-up Tensor Flow so that I can test this then.

**Comment by Cartucho** [comment]
I'm having some problems in trying to find a computer with GPU to test this haha

**Comment by vuthede** [comment]
Haha, I made a config file, which currently allow using CPU by default. We can download the model **ssdlite_mobilenet_v2_coco_2018_05_09** https://github.com/tensorflow/models/blob/master/research/object_detection/g3doc/detection_model_zoo.md and testing. It work quite ok on CPU i think.
At the moment, if the people using `Mask R-CNN`, the code still work ok with "bbox laleling". I haven't add pixel labeling as an option.

**Comment by Cartucho** [comment]
I will give it a try then!

**Comment by Cartucho** [comment]
I am getting this error:
image

**Comment by vuthede** [comment]
Ah sorry for no explanation Could u follow the instructions on Read me.

**Comment by vuthede** [comment]
@Cartucho I think it is because u may be unintentionally use python 2.x
Could you use python3.x by type **python3 main_auto.py** instead of **python main_auto.py**

**Comment by Cartucho** [comment]
Hello @vuthede ! I owe you an apology for not responding any sooner.
I have been moving out of my country since I just started a PhD in UK.
I think in April I will have time to merge this PR!

**Comment by vuthede** [comment]
@Cartucho Yeah no problem. Wish you have a good time on UKI

**Comment by lstappen** [comment]
Hi! 
when is the merge planned? Do you need any help for testing? Might be interesting to extend this functionality to an active learning component (update the model after every new label) and let it run in parallel.

**Comment by Cartucho** [comment]
@lstappen yes, do you think you can help testing it out?

**Comment by Cartucho** [comment]
@vuthede can you help me break this down file by file?

Let's start simple. For example, you are adding a people walking video. I think that is a great idea but if do that we should remove the video of the rabbit. Do you agree? You only need one video as a sample

**Comment by vuthede** [comment]
Yeah I think It is ok to keep rabbitor walking video. I am lean abit toward using people walking video because it may be more popular and practical to label.And it seems like there is no rabbit class on object detection.
I am very glad to help u, but could u show me **break this down file by side** means?

**Comment by Cartucho** [comment]
@vuthede it is essentially what we are doing now.

Ok, I will ask you to:

### Remove rabbit video

```
cd main/input
git rm video.mp4
```

and add modify the following line in `.gitignore`:

from: 
`main/input/video_mp4`

to:
`main/input/people_walking_mp4`


### Change the file `class_list.txt`

 from:

```
human
billiard ball
donut
rabbit
```

to 

```
person
billiard ball
donut
```

**Comment by vuthede** [comment]
Yeah! Done

File: main/release_output.sh [hunk 1]
```diff
@@ -0,0 +1,7 @@
+#!/usr/bin/env bash
+
+rm -rf output/.tracker/*
+rm -rf output/PASCAL_VOC/*
+rm -rf output/YOLO_darknet/*
+
+echo "Finish removing all outputs"
```
Comment (Cartucho) [inline main/release_output.sh]:
```suggestion
echo "Finished removing all the outputs"
```

**Comment by Cartucho** [comment]
ok, now I think that `release_output.sh` is not the best name, you suggest changing it to `remove_output.sh`:

```
cd main
git mv release_output.sh remove_output.sh
```

**Comment by Cartucho** [comment]
The file `object_detection/__pycache__/tf_object_detection.cpython-36.pyc` should not be here. You should solve it by adding the following line `*.pyc` to `.gitignore`

File: object_detection/.gitignore [hunk 2]
```diff
@@ -0,0 +1,2 @@
+faster_rcnn_nas_coco_2018_01_28
+ssdlite_mobilenet_v2_coco_2018_05_09
```
Comment (Cartucho) [inline object_detection/.gitignore]: There is no need to have multiple `.gitignore` files, let's use the one in the root of the directory only.

Also, all the models should be inside a folder: `object_detection/models` this way we can ignore that folder directly, so you can remove this file and add the following line to the root .gitignore `object_detection/models`

Comment (vuthede) [reply object_detection/.gitignore]: yeah it make sense. So i think i will
- remove the unnecessary .gitignore 
- create **object_detection/models** and put *models/* into  .gitignore
-  Fix the code so that it can work with new path of models

Comment (Cartucho) [reply object_detection/.gitignore]: Yes, that would be perfect!

Comment (vuthede) [reply object_detection/.gitignore]: ok

File: .gitignore [hunk 3]
```diff
@@ -1,4 +1,5 @@
 .idea/
 venv/
 main/input/people_walking_mp4
-main/output
\ No newline at end of file
+main/output
+models/
```
Comment (Cartucho) [inline .gitignore]:
```suggestion
object_detection/models
```

File: .gitignore [hunk 4]
```diff
@@ -1,4 +1,5 @@
 .idea/
 venv/
-main/input/video_mp4
-main/output
\ No newline at end of file
+main/input/people_walking_mp4
+main/output
+object_detection/models
```
Comment (Cartucho) [inline .gitignore]: add the following line:
`*.pyc`

to remove from this PR the following file:
`object_detection/__pycache__/tf_object_detection.cpython-36.pyc`

**Comment by Cartucho** [comment]
Great! There are 5 files left to be reviewed:

- [ ] `main/config.ini`
- [ ] `main/main_auto.py`
- [ ] `object_detection/tf_object_detection.py`
- [ ] `object_detection/utils.py`
- [ ] `README.md`

Relatively to `main/config.ini`. What do you are `OBJECT_IDS = 1,2` and `CUDA_VISIBLE_DEVICES = ''` standing for?

**Comment by vuthede** [comment]
- OBJECT_IDS = 1,2 that means the object detector will return only object which having objectId = 1(people) or 2. I believe that those objects ID are Ids used in COCO dataset,which are different than the classlist.txt we use
- CUDA_VISIBLE_DEVICES = ''" means using CPU. If there is GPU on our machine, we can define device like CUDA_VISIBLE_DEVICES = ''0" to use GPU 0

**Comment by Cartucho** [comment]
Ok, I still don't understand completely how the `OBJECT_ID` works, but I will have a look into your code to find out.

**Comment by vuthede** [comment]
thanks!

File: main/remove_output.sh [hunk 5]
```diff
@@ -0,0 +1,7 @@
+#!/usr/bin/env bash
+
+rm -rf output/.tracker/*
+rm -rf output/PASCAL_VOC/*
+rm -rf output/YOLO_darknet/*
```
Comment (Cartucho) [inline main/remove_output.sh]: couldn't all these lines be replaced by a single line `rm -rf output/*` ?

**Comment by Cartucho** [comment]
You also need to add `tensorflow` as one of the prerequisites in the README.

**Comment by Cartucho** [comment]
We don't really need the `config.ini` file since we are using `argsparse`.

The `main_auto.py` and `main.py` should be merged together and then one of the parameters in the `argsparse` should define whether or not the user wants to use the automatic labeling. For example, the default value could be False but the user could activate it to True.

What do you think?

Can I commit changes to the code too?

**Comment by vuthede** [comment]
yeah absolutely. You can change the code and commit it. Yeah I think it is a good idea to use only main.py and using argsparse

**Comment by Cartucho** [comment]
Hey, I will be merging it as is. It is a great feature congratulations! :+1: 

There is still some work to be done. Issues to be solved:

- [ ] the video is too big, needs to be cut, max 100 frames I would say
- [ ] the code should test if the model was indeed downloaded or not
- [ ] python3 and tensorflow are now a requirement. We should create a Conda environment
- [ ] does the code automatically tune to GPU vs. non-GPU usage?
- [ ] what happens if one runs the code twice? repeated labels?
- [ ] repeated code between `main.py` and `main_auto.py`
- [ ] remove `config.ini` and make the code use the classes in `class_list.txt`
- [ ] clean README it looks confusing now

These are the issues I plan to work on.
